### PR TITLE
Add hierarchical removal to the workflow and results pages

### PR DESCRIPTION
**Pull request by duncan-brown**
This pull requests adds hierarchical removal to the workflow and results pages.

The main open box page shows the results of all hierarchical removals (per @ahnitz request) since that is the final result.

Sec 7.01 will show no removals and be labelled **No Event Removal**

Sec 7.02 through 7.02 + ``max-hierarchical-removal`` will show each stage of hierarchical removal and will be labelled **Hierarchical Removal Level N**.

### CODE REVIEW

File: pycbc/workflow/plotting.py [hunk 1]
```diff
@@ -128,15 +128,25 @@ def make_throughput_plot(workflow, insp_files, out_dir, tags=None):
...
         tags = []
+
     makedir(out_dir)
     node = PlotExecutable(workflow.cp, 'page_foreground', ifos=workflow.ifos,
                     out_dir=out_dir, tags=tags).create_node()
     node.add_input_opt('--bank-file', bank_file)
     node.add_opt('--foreground-tag', ftag)
     node.add_input_opt('--trigger-file', trig_file)
+    if hierarchical_level: node.add_opt('--plot-hierarchical-level',
+                                        hierarchical_level)
```
Comment (stevereyes01) [inline pycbc/workflow/plotting.py]: Shouldn't this be `--use-hierarchical-level` or vice-versa?

Comment (duncan-brown) [reply pycbc/workflow/plotting.py]: fixed, thanks

**Comment by stevereyes01** [comment]
OOo, this will also fail because there's a line in each of the plotting codes to take `h_inc_back_num = args.hierarchical level`, those will have to be changed to `h_inc_back_num = args.use_hiearchical_level`.

After that it should be self-consistent.

**Comment by duncan-brown** [comment]
@stevereyes01 ok, fixed the command line arguments and the syntax error. Will test after Zoë is in bed.

**Comment by stevereyes01** [comment]
@duncan-brown Thanks. I'll be online for another 2 or 3 hours.

**Comment by stevereyes01** [comment]
This should be safe to merge as soon as I can get my patch in.

**Comment by stevereyes01** [comment]
Will merge when checks complete.

**Comment by duncan-brown** [comment]
@stevereyes01 hold on, I'm still debugging

**Comment by stevereyes01** [comment]
Waiting on Travis again.

**Comment by duncan-brown** [comment]
The setup of the results pages looks correct
```shell
(pycbc-dev)[dbrown@sugwg-osg results]$ ls -lR 7._open_box_result/
7._open_box_result/:
total 5
drwxr-x--x 2 dbrown lsc    3 Aug 25 19:59 7.01_no_event_removal
drwxr-x--x 2 dbrown lsc    3 Aug 25 20:07 7.02_hierarchical_removal_level_1
drwxr-x--x 2 dbrown lsc    3 Aug 25 20:00 7.03_hierarchical_removal_level_2
drwxr-x--x 2 dbrown lsc    3 Aug 25 20:01 7.04_hierarchical_removal_level_3
drwxr-x--x 2 dbrown lsc    3 Aug 25 20:01 7.05_hierarchical_removal_level_4
drwxr-x--x 2 dbrown lsc    3 Aug 25 20:02 7.06_hierarchical_removal_level_5
-rw-r----- 1 dbrown lsc 5605 Aug 25 19:59 well.html

7._open_box_result/7.01_no_event_removal:
total 2
-rw-r----- 1 dbrown lsc 9625 Aug 25 19:59 well.html

7._open_box_result/7.02_hierarchical_removal_level_1:
total 2
-rw-r----- 1 dbrown lsc 10465 Aug 25 20:00 well.html

7._open_box_result/7.03_hierarchical_removal_level_2:
total 2
-rw-r----- 1 dbrown lsc 10465 Aug 25 20:00 well.html

7._open_box_result/7.04_hierarchical_removal_level_3:
total 2
-rw-r----- 1 dbrown lsc 10465 Aug 25 20:01 well.html

7._open_box_result/7.05_hierarchical_removal_level_4:
total 2
-rw-r----- 1 dbrown lsc 10465 Aug 25 20:01 well.html

7._open_box_result/7.06_hierarchical_removal_level_5:
total 2
-rw-r----- 1 dbrown lsc 10465 Aug 25 20:02 well.html
```
and the content seems reasonable:
```html
<div class=pycbc-meta key="render-function" value="render_tmplt"></div><div class=pycbc-meta key="filenames" value="H1L1-PLOT_SNRIFAR_HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-1167523218-1584000.png,H1L1-PLOT_SNRRATEHIST_HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-1167523218-1584000.png,H1L1-PLOT_SNRIFAR_HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-1167523218-1584000.png,H1L1-PAGE_IFAR_HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-1167523218-1584000.png,H1L1-PAGE_FOREGROUND_HIERARCHICAL_LEVEL_01_HTML-1167523218-1584000.html"></div><!-- add a subtitle -->
```

**Comment by duncan-brown** [comment]
Workflow running now. If this completes successfully, then this is good to merge:

https://sugwg-osg.phy.syr.edu/pegasus/u/dbrown/r/496/w?wf_uuid=1d739c63-d862-4841-a44f-4c83e0c33ed8

https://sugwg-jobs.phy.syr.edu/~dbrown/osg/hm-test/o2-c01-bbh-rates-analysis-3-pre-1.7.9-hm/

**Comment by duncan-brown** [comment]
OK, looks good so far
```
14262278.0   dbrown                                                                                                             8/25 20:40   0+00:13:12 R  0       0.0 pegasus-dagman -p 0 -f -l . -Lockfile o2-c01-bbh-rates-analysis-3-
14262282.0    |-subdax_o2-c01-bbh-rates-analysis-3-pre-1.7.9-hm-test-main_ID0000001                                             8/25 20:43   0+00:09:51 R  10      0.3 condor_dagman -p 0 -f -l . -Notification never -Debug 3 -Lockfile 
14262291.0     |-plot_bank-H1L1_ID7_ID0000011                                                                                   8/25 20:47   0+00:02:40 R  1000    0.0 plot_bank-H1L1_ID7_ID0000011.sh
14262294.0     |-plot_spectrum-H1L1_ID101_ID0034649                                                                             8/25 20:47   0+00:02:40 R  1000    0.0 plot_spectrum-H1L1_ID101_ID0034649.sh
14262296.0     |-plot_binnedhist-DURATION_FULL_DATA-L1_ID78_ID0034626                                                           8/25 20:47   0+00:02:40 R  1000    0.0 plot_binnedhist-DURATION_FULL_DATA-L1_ID78_ID0034626.sh
14262297.0     |-plot_binnedhist-DURATION_FULL_DATA-H1_ID68_ID0034616                                                           8/25 20:47   0+00:02:40 R  1000    0.0 plot_binnedhist-DURATION_FULL_DATA-H1_ID68_ID0034616.sh
14262299.0     |-plot_range-BBH1-H1L1_ID104_ID0034652                                                                           8/25 20:47   0+00:02:40 R  1000    0.0 plot_range-BBH1-H1L1_ID104_ID0034652.sh
14262301.0     |-plot_range-BBH3-H1L1_ID106_ID0034654                                                                           8/25 20:47   0+00:02:40 R  1000    0.0 plot_range-BBH3-H1L1_ID106_ID0034654.sh
14262305.0     |-plot_range-SUMM-H1L1_ID102_ID0034650                                                                           8/25 20:47   0+00:02:31 R  1000    0.0 plot_range-SUMM-H1L1_ID102_ID0034650.sh
14262306.0     |-plot_range-NSBH-H1L1_ID103_ID0034651                                                                           8/25 20:47   0+00:02:40 R  1000    0.0 plot_range-NSBH-H1L1_ID103_ID0034651.sh
14262309.0     |-plot_range-BBH4-H1L1_ID107_ID0034655                                                                           8/25 20:47   0+00:02:19 R  1000    0.0 plot_range-BBH4-H1L1_ID107_ID0034655.sh
14262313.0     |-plot_range-BBH2-H1L1_ID105_ID0034653                                                                           8/25 20:53   0+00:00:27 R  1000    0.0 plot_range-BBH2-H1L1_ID105_ID0034653.sh
14262315.0     |-page_foreground-HIERARCHICAL_LEVEL_04_HTML-H1L1_ID51_ID0034599                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HIERARCHICAL_LEVEL_04_HTML-H1L1_ID51_ID0034599.sh
14262316.0     |-page_foreground-HIERARCHICAL_LEVEL_01_HTML-H1L1_ID36_ID0034584                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HIERARCHICAL_LEVEL_01_HTML-H1L1_ID36_ID0034584.sh
14262317.0     |-page_foreground-XMLLOUDEST-H1L1_ID26_ID0034574                                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-XMLLOUDEST-H1L1_ID26_ID0034574.sh
14262318.0     |-page_ifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID50_ID0034598  8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_12H_
14262319.0     |-plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID29_ID0034577                         8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR
14262320.0     |-page_ifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID40_ID0034588  8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_12H_
14262321.0     |-page_foreground-XMLALL-H1L1_ID25_ID0034573                                                                     8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-XMLALL-H1L1_ID25_ID0034573.sh
14262322.0     |-plot_snrifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID34_ID0034582   8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262323.0     |-plot_snrifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID52_ID0034600        8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262324.0     |-page_foreground-HIERARCHICAL_LEVEL_03_HTML-H1L1_ID46_ID0034594                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HIERARCHICAL_LEVEL_03_HTML-H1L1_ID46_ID0034594.sh
14262325.0     |-plot_snrratehist-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID33_ID0034581    8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_C
14262326.0     |-plot_snrratehist-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID53_ID0034601    8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_C
14262327.0     |-plot_snrratehist-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID28_ID0034576                          8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-
14262328.0     |-plot_snrratehist-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID38_ID0034586    8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_C
14262329.0     |-page_foreground-HTML-H1L1_ID24_ID0034572                                                                       8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HTML-H1L1_ID24_ID0034572.sh
14262330.0     |-plot_snrifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID32_ID0034580        8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262331.0     |-page_ifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID30_ID0034578                        8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BO
14262332.0     |-plot_snrifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID39_ID0034587   8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262333.0     |-plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID27_ID0034575                              8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1
14262334.0     |-plot_snrifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID54_ID0034602   8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262335.0     |-plot_snrifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID37_ID0034585        8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_02_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262336.0     |-plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID21_ID0034569                              8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1
14262337.0     |-page_ifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID45_ID0034593  8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_12H_
14262338.0     |-page_foreground-HTML-H1L1_ID31_ID0034579                                                                       8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HTML-H1L1_ID31_ID0034579.sh
14262339.0     |-page_ifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID35_ID0034583  8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-HIERARCHICAL_LEVEL_01_FULL_DATA_FULL_CUMULATIVE_CAT_12H_
14262340.0     |-plot_snrifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID44_ID0034592   8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262341.0     |-page_ifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_OPEN_BOX-H1L1_ID55_ID0034603  8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-HIERARCHICAL_LEVEL_05_FULL_DATA_FULL_CUMULATIVE_CAT_12H_
14262342.0     |-plot_snrifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID42_ID0034590        8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262343.0     |-plot_snrifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID47_ID0034595        8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262344.0     |-plot_snrratehist-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID22_ID0034570                          8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-
14262345.0     |-plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_CLOSED-H1L1_ID20_ID0034568                       8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_CLOS
14262346.0     |-plot_snrratehist-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID43_ID0034591    8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-HIERARCHICAL_LEVEL_03_FULL_DATA_FULL_CUMULATIVE_C
14262347.0     |-plot_snrifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_IFAR-H1L1_ID49_ID0034597   8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrifar-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_1
14262348.0     |-page_foreground-HIERARCHICAL_LEVEL_05_HTML-H1L1_ID56_ID0034604                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HIERARCHICAL_LEVEL_05_HTML-H1L1_ID56_ID0034604.sh
14262349.0     |-page_foreground-HIERARCHICAL_LEVEL_02_HTML-H1L1_ID41_ID0034589                                                 8/25 20:53   0+00:00:00 I  1000    0.0 page_foreground-HIERARCHICAL_LEVEL_02_HTML-H1L1_ID41_ID0034589.sh
14262350.0     |-page_ifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_CLOSED_BOX-H1L1_ID23_ID0034571                      8/25 20:53   0+00:00:00 I  1000    0.0 page_ifar-FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL_CLOSED_
14262351.0     |-plot_snrratehist-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-H1L1_ID48_ID0034596    8/25 20:53   0+00:00:00 I  1000    0.0 plot_snrratehist-HIERARCHICAL_LEVEL_04_FULL_DATA_FULL_CUMULATIVE_C
14262352.0     |-foreground_minifollowup-H1L1_ID108_ID0034656                                                                   8/25 20:53   0+00:00:00 I  500     0.0 foreground_minifollowup-H1L1_ID108_ID0034656.sh
14262353.0     |-statmap_inj-LNMINJ_INJ-H1L1_ID178_ID0052020                                                                    8/25 20:53   0+00:00:00 I  500     0.0 statmap_inj-LNMINJ_INJ-H1L1_ID178_ID0052020.sh
14262354.0     |-foreground_censor-H1L1_ID19_ID0034567                                                                          8/25 20:53   0+00:00:00 I  500     0.0 foreground_censor-H1L1_ID19_ID0034567.sh
14262355.0     |-statmap_inj-IMFINJ_INJ-H1L1_ID136_ID0043335                                                                    8/25 20:53   0+00:00:00 I  500     0.0 statmap_inj-IMFINJ_INJ-H1L1_ID136_ID0043335.sh
```

**Comment by duncan-brown** [comment]
@stevereyes01 can you confirm that ``page_foreground-XMLLOUDEST`` and ``page_foreground-XMLALL`` show the full removed results (same as ``page_foreground-HTML``)? I think that is the case, as they are run with no arguments.

**Comment by stevereyes01** [comment]
Ooo, wow, looks really neat! I really appreciate all of your help Duncan.

**Comment by stevereyes01** [comment]
@duncan-brown The XML files will have the full removed results and every stage of hierarchical removal correct (if given with that option), except the hierarchical removal xml files won't be populated with an exclusive IFAR nor exclusive FAP. That's really not the end of the world in my opinion since that won't change across any levels, it's just fancy bookkeeping.

**Pull request merged**